crates/kornia-imgproc/src/filter/ops.rs

```diff
@@ -118,6 +122,535 @@ pub fn box_blur_fast<const C: usize>(
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dx.cols(),
+            dx.rows(),
+        ));
+    }
+
+    if src.size() != dy.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dy.cols(),
+            dy.rows(),
+        ));
+    }
+
+    let (sobel_x, sobel_y) = kernels::normalized_sobel_kernel3();
+    let cols = src.cols();
+
+    let src_data = src.as_slice();
+
+    dx.as_slice_mut()
+        .chunks_mut(cols * C)
+        .zip(dy.as_slice_mut().chunks_mut(cols * C))
+        .enumerate()
+        .for_each(|(r, (dx_row, dy_row))| {
+            dx_row
+                .chunks_mut(C)
+                .zip(dy_row.chunks_mut(C))
+                .enumerate()
+                .for_each(|(c, (dx_c, dy_c))| {
+                    let mut sum_x = [0.0; C];
+                    let mut sum_y = [0.0; C];
+                    for dy in 0..3 {
+                        for dx in 0..3 {
+                            let row = (r + dy).min(src.rows()).max(1) - 1;
+                            let col = (c + dx).min(src.cols()).max(1) - 1;
+                            for ch in 0..C {
+                                let src_pix_offset = (row * src.cols() + col) * C + ch;
+                                let val = unsafe { src_data.get_unchecked(src_pix_offset) };
+                                sum_x[ch] += val * sobel_x[dy][dx];
+                                sum_y[ch] += val * sobel_y[dy][dx];
+                            }
+                        }
+                    }
+                    dx_c.copy_from_slice(&sum_x);
+                    dy_c.copy_from_slice(&sum_y);
+                });
+        });
+
+    Ok(())
+}
+
+/// Compute the first order image derivative in both x and y using a Sobel operator.
+///
+/// # Arguments
+///
+/// * `src` - The source image with shape (H, W).
+/// * `dst` - The destination image with shape (H, W, 2).
+pub fn spatial_gradient_float_rayon_parallel<const C: usize>(
+    src: &Image<f32, C>,
+    dx: &mut Image<f32, C>,
+    dy: &mut Image<f32, C>,
+) -> Result<(), ImageError> {
+    if src.size() != dx.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dx.cols(),
+            dx.rows(),
+        ));
+    }
+
+    if src.size() != dy.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dy.cols(),
+            dy.rows(),
+        ));
+    }
+
+    let (sobel_x, sobel_y) = kernels::normalized_sobel_kernel3();
+    let cols = src.cols();
+
+    let src_data = src.as_slice();
+
+    dx.as_slice_mut()
+        .par_chunks_mut(cols * C)
+        .zip(dy.as_slice_mut().par_chunks_mut(cols * C))
+        .enumerate()
+        .for_each(|(r, (dx_row, dy_row))| {
+            dx_row
+                .par_chunks_mut(C)
```

Comment:
```suggestion
                .chunks_mut(C)
```
have you explored this option just parallelization over the row (not channels) ?

I would consider only:
- serial (raw iterators)
- par by rows
- par by rows / par by channels

if you want to skip borders just do (or similar) be
```rust
if r == 0 || r == cols-1 {
   return;
}
```

Reply:
this function should be very simple, do not over-complicate the code, and please clean unused code